File: system-test/typescript/src/testkit/driver/adapter/external/clock/clock-real-driver.ts
import { Result, success } from '../../../../common/result.js';
import { ClockErrorResponse, GetTimeResponse, ReturnsTimeRequest } from '../../../../common/dtos.js';
import { ClockDriver } from '../../../port/external/clock/clock-driver.js';

export class ClockRealDriver implements ClockDriver {
  async goToClock(): Promise<Result<void, ClockErrorResponse>> {
    return success(undefined);
  }

  async getTime(): Promise<Result<GetTimeResponse, ClockErrorResponse>> {
    return success({ time: new Date().toISOString() });
  }

  async returnsTime(_request: ReturnsTimeRequest): Promise<Result<void, ClockErrorResponse>> {
    return success(undefined);
  }

  async close(): Promise<void> {}
}

File: system-test/java/src/main/java/com/mycompany/myshop/testkit/driver/adapter/external/clock/ClockRealDriver.java
package com.mycompany.myshop.testkit.driver.adapter.external.clock;

import com.mycompany.myshop.testkit.driver.port.external.clock.ClockDriver;

import com.mycompany.myshop.testkit.driver.adapter.external.clock.client.ClockRealClient;
import com.mycompany.myshop.testkit.driver.port.external.clock.dtos.GetTimeResponse;
import com.mycompany.myshop.testkit.driver.port.external.clock.dtos.ReturnsTimeRequest;
import com.mycompany.myshop.testkit.driver.port.external.clock.dtos.error.ClockErrorResponse;
import com.mycompany.myshop.testkit.common.Result;

public class ClockRealDriver implements ClockDriver {
    private final ClockRealClient client;

    public ClockRealDriver() {
        this.client = new ClockRealClient();
    }

    @Override
    public void close() {
        // No resources to close - ClockRealClient has no HTTP or other resources
    }

    @Override
    public Result<Void, ClockErrorResponse> goToClock() {
        return client.checkHealth()
                .mapError(ext -> new ClockErrorResponse(ext.getMessage()));
    }

    @Override
    public Result<GetTimeResponse, ClockErrorResponse> getTime() {
        return client.getTime()
                .map(ext -> GetTimeResponse.builder().time(ext.getTime()).build())
                .mapError(ext -> new ClockErrorResponse(ext.getMessage()));
    }

    @Override
    public Result<Void, ClockErrorResponse> returnsTime(ReturnsTimeRequest request) {
        // No-op because real clock cannot be configured
        return Result.success();
    }
}
